Backport H5Dchunk_iter to 1.12 branch

mkitti opened the pull request:
H5Dchunk_iter was reintegrated reverting #733.

Incorporates changes to `H5chunk_iter` from
* #161
* #722
* #733
* #1474
* #1820 
* #1969

To do:
- [x] Backport #1969 
xref: #1968 (for 1.10 branch)

derobins committed:
Backport to 1.12: Adds a quick for for some egregious chunk_info badness (#722)

derobins committed:
Backport to 1.12: Converts testhdf5 macros to h5test macros in chunk_info.c (#1820)

The two macro schemes were not designed to work together. Also
quiets some MSVC warnings about comparing pointers and integers.

mkitti committed:
Revert "Revert H5Dchunk_iter changes in hdf5_1_12_1 (#733)"

This reverts commit 10abe9a8b4985a18a0c88615686184d233d7b845.

mkitti committed:
Apply clang-format

mkitti committed:
Reincorporate spelling fix from #1166

mkitti committed:
Incorporate H5Dchunk_iter changes from #161

derobins commented on src/H5VLnative_dataset.c:
```diff
@@ -346,6 +346,31 @@ H5VL__native_dataset_specific(void *obj, H5VL_dataset_specific_t specific_type,
+             *      operations, so this is a no-op.
+             */
+            break;
+        }
```
The dataset wait case doesn't belong in 1.12

mkitti replied:
Resolved in [0b97a78](https://github.com/HDFGroup/hdf5/pull/1970/commits/0b97a78647db222e91d0b9a466f5abf8392fb100)

We are now using `H5VL_NATIVE_DATASET_CHUNK_ITER` instead.

derobins commented on src/H5VLconnector.h:
```diff
@@ -81,7 +81,9 @@ typedef enum H5VL_dataset_get_t {
     H5VL_DATASET_FLUSH,      /* H5Dflush                            */
-    H5VL_DATASET_REFRESH     /* H5Drefresh                          */
+    H5VL_DATASET_REFRESH,    /* H5Drefresh                          */
+    H5VL_DATASET_WAIT,       /* H5Dwait                             */
```
Dataset wait doesn't belong in 1.12

mkitti replied:
Resolved in [0b97a78](https://github.com/HDFGroup/hdf5/pull/1970/commits/0b97a78647db222e91d0b9a466f5abf8392fb100)

We are now using `H5VL_NATIVE_DATASET_CHUNK_ITER` instead.

mkitti committed:
Backport to 1.12: H5Dchunk_iter now passes offsets in units of dataset elements, fix #1419 (#1969)

* H5Dchunk_iter now passes chunk dimension scaled offsets, fix #1419

* Update docs for H5Dchunk_iter, H5Dget_chunk_info*

Modified description for `H5Dchunk_iter`, `H5Dget_chunk_info`, and `H5Dget_chunk_info_by_coord` to the following

 * offset          Logical position of the chunk’s first element in units of dataset elements
 * filter_mask  Bitmask indicating the filters used when the chunk was written
 * size             Chunk size in bytes, 0 if the chunk does not exist

mkitti committed:
Sync H5Dchunk_iter documentation with develop

mkitti committed:
Remove H5VL_DATASET_WAIT references from 1.12

mkitti committed:
Backport to 1.12 #161, #1474, review comments

jhendersonHDF reviewed (approved):
LGTM. Thanks for the work in backporting this!
